core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -415,6 +421,25 @@ public void updateShardState(final ShardRouting newRouting,
                 }
 
                 changeState(IndexShardState.STARTED, "global state is [" + newRouting.state() + "]");
+
+                final Runnable adjustBreaker = () -> {
+                    CircuitBreaker cb = circuitBreakerService.getBreaker(CircuitBreaker.ACCOUNTING);
+                    final SegmentsStats ss = segmentStats(false);
+                    final long newMemory = ss.getMemoryInBytes();
+                    final long currentMemory = memoryAccountingBytes.get();
+                    if (newMemory != currentMemory) {
+                        // Being paranoid about multiple refresh listeners executing at a time here
```

Comment:
I figured the tradeoff of potentially being behind for one refresh was better than having locking/spinning during every refresh. That, and I'm pretty sure refreshes don't happen in parallel.

Do you think it's worth doing the spinning/locking rather than being behind for a refresh?

Reply:
Refreshes do not but these listeners from subsequent refreshes could race with each other, no?

Reply:
Let's assume 2 operations, t1 and t2

- t1 starts refreshing, gets currentMemory=0 newMemory=10
- t2 starts refreshing, gets currentMemory=0 newMemory=20
- t1 successfully updates, now currentMemory=10
- t2 fails the `compareAndSet`

I think that explains what you are talking about right?

I think this is okay though? because next time a refresh happens it will get currentMemory correctly and change the memory to be correct.

I erred on the side of low impact and eventual correctness rather than higher impact with spinning/locking and always correctness.

Do you think it's better to be on the higher impact but more correct side? I think it's unlikely that refreshes will be concurrent, but I'm totally up for changing it if you think it will be a large impact.

Reply:
Yes, that's exactly what I'm talking about. I don't think it's that much higher impact to be correct here, so that would be my preference.

Reply:
Okay, I changed it in 7804edbdb0a13506abed04d4fb73deb4a997499b

Reply:
Please don’t synchronize on this, that’s dangerous as someone else that has a reference to the index shard can synchronize on it.

Reply:
I changed it to synchronize on something different, thanks for the suggestion